File: tests/test_api.py
from datetime import datetime, timedelta

from fastapi.testclient import TestClient
import pytest

from database import FINE_PER_DAY, LOAN_DAYS, Issue, calculate_fine, days_overdue
from main import app


@pytest.fixture(scope="module")
def client():
    with TestClient(app) as test_client:
        yield test_client


def auth_headers(client):
    response = client.post("/api/login", json={"username": "admin", "password": "admin123"})
    assert response.status_code == 200
    token = response.json()["token"]
    return {"Authorization": f"Bearer {token}"}


# ==============================================================================
# FR-01: User Login & Authentication
# ==============================================================================

def test_login_success_and_me(client):
    response = client.post("/api/login", json={"username": "admin", "password": "admin123"})
    assert response.status_code == 200
    data = response.json()
    assert "token" in data
    assert data["username"] == "admin"

    headers = {"Authorization": f"Bearer {data['token']}"}
    me = client.get("/api/me", headers=headers)
    assert me.status_code == 200
    assert me.json()["username"] == "admin"


def test_login_failures(client):
    # Invalid password
    bad_pass = client.post("/api/login", json={"username": "admin", "password": "wrongpassword"})
    assert bad_pass.status_code == 401

    # Nonexistent user
    bad_user = client.post("/api/login", json={"username": "nobody", "password": "admin123"})
    assert bad_user.status_code == 401

    # Unauthenticated endpoints
    assert client.get("/api/books").status_code == 401
    assert client.get("/api/dashboard").status_code == 401
    assert client.get("/api/fines").status_code == 401
    assert client.get("/api/availability").status_code == 401


# ==============================================================================
# FR-02: Book Management (Add, Update, Delete, Search)
# ==============================================================================

def test_book_crud_lifecycle(client):
    headers = auth_headers(client)

    # Add book (code generated automatically, status="available")
    created = client.post(
        "/api/books",
        json={"title": "Introduction to Algorithms", "author": "Thomas H. Cormen"},
        headers=headers,
    )
    assert created.status_code == 201
    book = created.json()
    assert book["title"] == "Introduction to Algorithms"
    assert book["author"] == "Thomas H. Cormen"
    assert book["status"] == "available"
    assert book["code"].startswith("B")
    book_id = book["id"]
    book_code = book["code"]

    # Search by title
    search_title = client.get("/api/books", params={"q": "Algorithms"}, headers=headers)
    assert search_title.status_code == 200
    assert any(b["id"] == book_id for b in search_title.json())

    # Search by author
    search_author = client.get("/api/books", params={"q": "Cormen"}, headers=headers)
    assert search_author.status_code == 200
    assert any(b["id"] == book_id for b in search_author.json())

    # Search by code
    search_code = client.get("/api/books", params={"q": book_code}, headers=headers)
    assert search_code.status_code == 200
    assert any(b["id"] == book_id for b in search_code.json())

    # Sorting
    for sort_key in ["title", "author", "status", "default"]:
        sort_res = client.get("/api/books", params={"sort": sort_key}, headers=headers)
        assert sort_res.status_code == 200
        assert len(sort_res.json()) > 0

    # Update book
    updated = client.put(
        f"/api/books/{book_id}",
        json={"title": "Introduction to Algorithms, 4th Ed.", "author": "Thomas H. Cormen"},
        headers=headers,
    )
    assert updated.status_code == 200
    assert updated.json()["title"] == "Introduction to Algorithms, 4th Ed."

    # Delete book (available with no issue history)
    deleted = client.delete(f"/api/books/{book_id}", headers=headers)
    assert deleted.status_code == 200
    assert deleted.json() == {"ok": True}

    # Verify deleted
    after_delete = client.get("/api/books", params={"q": book_code}, headers=headers)
    assert not any(b["id"] == book_id for b in after_delete.json())


def test_book_validation_errors(client):
    headers = auth_headers(client)

    # Empty / whitespace title or author rejected with 400
    empty_title = client.post("/api/books", json={"title": "   ", "author": "Valid Author"}, headers=headers)
    assert empty_title.status_code == 400

    empty_author = client.post("/api/books", json={"title": "Valid Title", "author": "   "}, headers=headers)
    assert empty_author.status_code == 400

    # Nonexistent book update / delete returns 404
    assert client.put("/api/books/99999", json={"title": "T", "author": "A"}, headers=headers).status_code == 404
    assert client.delete("/api/books/99999", headers=headers).status_code == 404


def test_book_delete_restrictions(client):
    headers = auth_headers(client)

    # Create a book
    res = client.post("/api/books", json={"title": "Restricted Book", "author": "Author R"}, headers=headers)
    book_id = res.json()["id"]

    # Issue book -> status is now "borrowed"
    client.post("/api/issues", json={"book_id": book_id, "borrower": "Borrower 1"}, headers=headers)

    # Rule: Borrowed books cannot be deleted
    del_borrowed = client.delete(f"/api/books/{book_id}", headers=headers)
    assert del_borrowed.status_code == 400
    assert "borrowed or overdue" in del_borrowed.json()["detail"]

    # Return book
    client.post("/api/returns", json={"book_id": book_id}, headers=headers)

    # Rule: Books with circulation history cannot be deleted
    del_history = client.delete(f"/api/books/{book_id}", headers=headers)
    assert del_history.status_code == 400
    assert "issue history" in del_history.json()["detail"]


# ==============================================================================
# FR-03: Book Issue & Loan Tracking
# ==============================================================================

def test_book_issue_lifecycle(client):
    headers = auth_headers(client)

    # Create book
    book = client.post("/api/books", json={"title": "Issue Test Book", "author": "Author I"}, headers=headers).json()
    book_id = book["id"]

    # Issue book (loan period is 14 days)
    issued = client.post(
        "/api/issues",
        json={"book_id": book_id, "borrower": "Student X"},
        headers=headers,
    )
    assert issued.status_code == 200
    issue_data = issued.json()
    assert issue_data["borrower"] == "Student X"
    assert issue_data["status"] == "borrowed"
    assert issue_data["book_id"] == book_id

    # Check due date is ~14 days from issue date
    issue_dt = datetime.fromisoformat(issue_data["issued_at"])
    due_dt = datetime.fromisoformat(issue_data["due_at"])
    assert (due_dt - issue_dt).days == LOAN_DAYS

    # Rule: Cannot issue already borrowed book
    second_issue = client.post("/api/issues", json={"book_id": book_id, "borrower": "Student Y"}, headers=headers)
    assert second_issue.status_code == 400

    # Rule: Cannot issue with empty borrower name
    empty_borrower = client.post("/api/issues", json={"book_id": book_id, "borrower": "   "}, headers=headers)
    assert empty_borrower.status_code == 400

    # Rule: Nonexistent book issue returns 404
    assert client.post("/api/issues", json={"book_id": 99999, "borrower": "Student Z"}, headers=headers).status_code == 404

    # Active issues listing
    issues_list = client.get("/api/issues", params={"active": True}, headers=headers)
    assert issues_list.status_code == 200
    assert any(item["id"] == issue_data["id"] for item in issues_list.json())


# ==============================================================================
# FR-04: Book Return & Availability Restoration
# ==============================================================================

def test_book_return_lifecycle(client):
    headers = auth_headers(client)

    # Create and issue book
    book = client.post("/api/books", json={"title": "Return Test Book", "author": "Author R"}, headers=headers).json()
    book_id = book["id"]
    client.post("/api/issues", json={"book_id": book_id, "borrower": "Student Return"}, headers=headers)

    # Return book
    ret_res = client.post("/api/returns", json={"book_id": book_id}, headers=headers)
    assert ret_res.status_code == 200
    ret_data = ret_res.json()
    assert ret_data["status"] == "returned"
    assert ret_data["returned_at"] is not None

    # Verify book status restored to "available"
    books_res = client.get("/api/books", params={"q": "Return Test Book"}, headers=headers)
    matched = [b for b in books_res.json() if b["id"] == book_id]
    assert len(matched) == 1
    assert matched[0]["status"] == "available"

    # Rule: Cannot return an unborrowed book
    repeat_return = client.post("/api/returns", json={"book_id": book_id}, headers=headers)
    assert repeat_return.status_code == 400

    # Rule: Nonexistent book return returns 404
    assert client.post("/api/returns", json={"book_id": 99999}, headers=headers).status_code == 404

    # Returned items listing
    returns_list = client.get("/api/returns", headers=headers)
    assert returns_list.status_code == 200
    assert any(item["book_id"] == book_id for item in returns_list.json())


# ==============================================================================
# FR-05: Fine Calculation & Management
# ==============================================================================

def test_fine_calculation_helpers():
    due = datetime(2026, 1, 1, 12, 0, 0)

    # On or before due date: 0 fine
    issue_on_time = Issue(book_id=1, borrower="T", issued_at=due - timedelta(days=14), due_at=due)
    assert days_overdue(issue_on_time, due) == 0
    assert calculate_fine(issue_on_time, due) == 0.0

    # 5 days overdue: 5 * 10 = ₹50
    now = due + timedelta(days=5)
    assert days_overdue(issue_on_time, now) == 5
    assert calculate_fine(issue_on_time, now) == 5 * FINE_PER_DAY


def test_fines_endpoint(client):
    headers = auth_headers(client)

    fines_res = client.get("/api/fines", headers=headers)
    assert fines_res.status_code == 200
    data = fines_res.json()
    assert "total" in data
    assert "count" in data
    assert "items" in data
    assert isinstance(data["items"], list)
    if data["items"]:
        # Should be ordered descending by fine amount
        fines = [item["fine"] for item in data["items"]]
        assert fines == sorted(fines, reverse=True)


# ==============================================================================
# FR-06: Book Availability
# ==============================================================================

def test_book_availability_filters(client):
    headers = auth_headers(client)

    # All books
    all_books = client.get("/api/availability", params={"status": "all"}, headers=headers)
    assert all_books.status_code == 200
    assert len(all_books.json()) > 0

    # Available filter
    avail_books = client.get("/api/availability", params={"status": "available"}, headers=headers)
    assert avail_books.status_code == 200
    assert all(b["status"] == "available" for b in avail_books.json())

    # Borrowed filter (includes borrowed and overdue)
    borrowed_books = client.get("/api/availability", params={"status": "borrowed"}, headers=headers)
    assert borrowed_books.status_code == 200
    assert all(b["status"] in ("borrowed", "overdue") for b in borrowed_books.json())

    # Overdue filter
    overdue_books = client.get("/api/availability", params={"status": "overdue"}, headers=headers)
    assert overdue_books.status_code == 200
    assert all(b["status"] == "overdue" for b in overdue_books.json())


# ==============================================================================
# Dashboard & Reports
# ==============================================================================

def test_dashboard_metrics_and_reports(client):
    headers = auth_headers(client)

    dashboard = client.get("/api/dashboard", headers=headers)
    assert dashboard.status_code == 200
    data = dashboard.json()

    # Verify all expected metric cards and reports are present
    required_keys = [
        "total_books",
        "available",
        "borrowed",
        "overdue",
        "active_loans",
        "outstanding_fines",
        "collected_fines",
        "recent_issues",
        "overdue_items",
    ]
    for key in required_keys:
        assert key in data, f"Missing dashboard key: {key}"

    assert data["total_books"] >= 0
    assert data["available"] >= 0
    assert data["borrowed"] >= 0
    assert data["overdue"] >= 0
    assert data["outstanding_fines"] >= 0.0
    assert data["collected_fines"] >= 0.0
    assert isinstance(data["recent_issues"], list)
    assert isinstance(data["overdue_items"], list)


# ==============================================================================
# Static Assets & Health Check
# ==============================================================================

def test_static_and_health_endpoints(client):
    assert client.get("/health").status_code == 200

    # Favicon returns 204 No Content
    fav = client.get("/favicon.ico")
    assert fav.status_code == 204

    # HTML home
    home = client.get("/")
    assert home.status_code == 200
    assert "LIBRARY MANAGEMENT SYSTEM" in home.text or "Library Management System" in home.text

    # CSS
    css = client.get("/style.css")
    assert css.status_code == 200
    assert "text/css" in css.headers["content-type"]

    # JS
    js = client.get("/script.js")
    assert js.status_code == 200
    assert "text/javascript" in js.headers["content-type"]

File: tests/test_api.js
import assert from "assert/strict";

const BASE_URL = "http://127.0.0.1:3000";

async function runTests() {
  console.log("=== Starting LIBRA End-to-End API Test Suite ===");
  let authToken = "";

  // 1. Health check
  console.log("\n[Test 1] Health Check...");
  const healthRes = await fetch(`${BASE_URL}/health`);
  assert.equal(healthRes.status, 200);
  const healthData = await healthRes.json();
  assert.equal(healthData.status, "ok");
  console.log("✓ Health Check passed:", healthData);

  // 2. FR-01: Authentication with valid credentials
  console.log("\n[Test 2] FR-01: Login with valid credentials (admin / admin123)...");
  const loginRes = await fetch(`${BASE_URL}/api/login`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ username: "admin", password: "admin123" }),
  });
  assert.equal(loginRes.status, 200);
  const loginData = await loginRes.json();
  assert.ok(loginData.token);
  assert.equal(loginData.username, "admin");
  authToken = loginData.token;
  console.log("✓ Valid login successful, token acquired.");

  // 3. FR-01: Authentication rejection with invalid credentials
  console.log("\n[Test 3] FR-01: Login with invalid credentials...");
  const badLoginRes = await fetch(`${BASE_URL}/api/login`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ username: "admin", password: "wrongpassword" }),
  });
  assert.equal(badLoginRes.status, 401);
  console.log("✓ Bad login rejected with 401 Unauthorized.");

  // 3a. FR-01: User Registration
  console.log("\n[Test 3a] User Registration (Create New Account)...");
  const testUser = `newuser_${Date.now()}`;
  const regRes = await fetch(`${BASE_URL}/api/register`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ username: testUser, password: "password123" }),
  });
  assert.equal(regRes.status, 201);
  const regData = await regRes.json();
  assert.ok(regData.token);
  assert.equal(regData.username, testUser);
  console.log("✓ New user registered successfully:", testUser);

  // 3b. Duplicate registration rejection
  console.log("\n[Test 3b] Duplicate Registration Rejection...");
  const dupRes = await fetch(`${BASE_URL}/api/register`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ username: testUser, password: "password123" }),
  });
  assert.equal(dupRes.status, 409);
  console.log("✓ Duplicate username rejected with 409 Conflict.");

  // 3c. Login with newly created credentials
  console.log("\n[Test 3c] Login with newly created user credentials...");
  const newLoginRes = await fetch(`${BASE_URL}/api/login`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ username: testUser, password: "password123" }),
  });
  assert.equal(newLoginRes.status, 200);
  const newLoginData = await newLoginRes.json();
  assert.equal(newLoginData.username, testUser);
  console.log("✓ Successfully logged in with newly created account credentials.");

  // 4. Auth headers helper
  const authHeaders = {
    "Content-Type": "application/json",
    Authorization: `Bearer ${authToken}`,
  };

  // 5. FR-01: Check /api/me
  console.log("\n[Test 4] Check /api/me session...");
  const meRes = await fetch(`${BASE_URL}/api/me`, { headers: authHeaders });
  assert.equal(meRes.status, 200);
  const meData = await meRes.json();
  assert.equal(meData.username, "admin");
  console.log("✓ Authenticated user profile matches.");

  // 6. Dashboard metrics
  console.log("\n[Test 5] Dashboard metrics...");
  const dashRes = await fetch(`${BASE_URL}/api/dashboard`, { headers: authHeaders });
  assert.equal(dashRes.status, 200);
  const dash = await dashRes.json();
  assert.ok(typeof dash.total_books === "number");
  assert.ok(typeof dash.available === "number");
  assert.ok(typeof dash.borrowed === "number");
  assert.ok(typeof dash.overdue === "number");
  assert.ok(Array.isArray(dash.recent_issues));
  console.log("✓ Dashboard metrics verified. Total books:", dash.total_books);

  // 7. FR-02: Get books catalog and search
  console.log("\n[Test 6] FR-02: Books Catalog & Search...");
  const booksRes = await fetch(`${BASE_URL}/api/books`, { headers: authHeaders });
  assert.equal(booksRes.status, 200);
  const allBooks = await booksRes.json();
  assert.ok(allBooks.length >= 8);

  const searchRes = await fetch(`${BASE_URL}/api/books?q=Python`, { headers: authHeaders });
  assert.equal(searchRes.status, 200);
  const searchResults = await searchRes.json();
  assert.ok(searchResults.some((b) => b.title.includes("Python")));
  console.log("✓ Books catalog & search verified.");

  // 8. FR-02: Add new book
  console.log("\n[Test 7] FR-02: Add new book...");
  const newBookRes = await fetch(`${BASE_URL}/api/books`, {
    method: "POST",
    headers: authHeaders,
    body: JSON.stringify({
      title: "Design Patterns in Practice",
      author: "Erich Gamma",
    }),
  });
  assert.equal(newBookRes.status, 201);
  const newBook = await newBookRes.json();
  assert.ok(newBook.id);
  assert.ok(newBook.code);
  assert.equal(newBook.status, "available");
  assert.equal(newBook.title, "Design Patterns in Practice");
  console.log("✓ New book added with code:", newBook.code, "ID:", newBook.id);

  // 9. FR-02: Update book
  console.log("\n[Test 8] FR-02: Update book...");
  const updateRes = await fetch(`${BASE_URL}/api/books/${newBook.id}`, {
    method: "PUT",
    headers: authHeaders,
    body: JSON.stringify({
      title: "Design Patterns: Elements of Reusable Object-Oriented Software",
      author: "Gang of Four",
    }),
  });
  assert.equal(updateRes.status, 200);
  const updatedBook = await updateRes.json();
  assert.equal(updatedBook.title, "Design Patterns: Elements of Reusable Object-Oriented Software");
  assert.equal(updatedBook.author, "Gang of Four");
  console.log("✓ Book updated successfully.");

  // 10. FR-03: Issue the book
  console.log("\n[Test 9] FR-03: Issue book to borrower...");
  const issueRes = await fetch(`${BASE_URL}/api/issues`, {
    method: "POST",
    headers: authHeaders,
    body: JSON.stringify({
      book_id: newBook.id,
      borrower: "Alice Smith",
    }),
  });
  assert.equal(issueRes.status, 200);
  const issueData = await issueRes.json();
  assert.equal(issueData.book_id, newBook.id);
  assert.equal(issueData.borrower, "Alice Smith");
  assert.equal(issueData.status, "borrowed");
  console.log("✓ Book issued successfully to Alice Smith.");

  // 11. Operational rule: Cannot delete borrowed book
  console.log("\n[Test 10] Operational Rule: Cannot delete borrowed book...");
  const delBorrowedRes = await fetch(`${BASE_URL}/api/books/${newBook.id}`, {
    method: "DELETE",
    headers: authHeaders,
  });
  assert.equal(delBorrowedRes.status, 400);
  console.log("✓ Deleting borrowed book correctly rejected with 400.");

  // 12. FR-04: Return the book
  console.log("\n[Test 11] FR-04: Return book...");
  const returnRes = await fetch(`${BASE_URL}/api/returns`, {
    method: "POST",
    headers: authHeaders,
    body: JSON.stringify({ book_id: newBook.id }),
  });
  assert.equal(returnRes.status, 200);
  const returnData = await returnRes.json();
  assert.equal(returnData.book_id, newBook.id);
  assert.equal(returnData.status, "returned");
  console.log("✓ Book returned successfully.");

  // 13. Operational rule: Cannot delete book with circulation history
  console.log("\n[Test 12] Operational Rule: Cannot delete book with circulation history...");
  const delHistoryRes = await fetch(`${BASE_URL}/api/books/${newBook.id}`, {
    method: "DELETE",
    headers: authHeaders,
  });
  assert.equal(delHistoryRes.status, 400);
  console.log("✓ Deleting book with history correctly rejected with 400.");

  // 14. FR-05: Fines management
  console.log("\n[Test 13] FR-05: Fines management calculation...");
  const finesRes = await fetch(`${BASE_URL}/api/fines`, { headers: authHeaders });
  assert.equal(finesRes.status, 200);
  const finesData = await finesRes.json();
  assert.ok(typeof finesData.total === "number");
  assert.ok(Array.isArray(finesData.items));
  console.log("✓ Fines report verified. Active fine records:", finesData.count, "Total fine: ₹" + finesData.total);

  // 15. FR-06: Availability filter
  console.log("\n[Test 14] FR-06: Availability catalog...");
  const availRes = await fetch(`${BASE_URL}/api/availability?status=available`, { headers: authHeaders });
  assert.equal(availRes.status, 200);
  const availableBooks = await availRes.json();
  assert.ok(availableBooks.every((b) => b.status === "available"));
  console.log("✓ Availability filtering verified. Available books:", availableBooks.length);

  // 16. Uncirculated book deletion
  console.log("\n[Test 15] Operational Rule: Deleting an uncirculated book...");
  const tempBookRes = await fetch(`${BASE_URL}/api/books`, {
    method: "POST",
    headers: authHeaders,
    body: JSON.stringify({ title: "Temporary Book", author: "Anon" }),
  });
  const tempBook = await tempBookRes.json();
  const delTempRes = await fetch(`${BASE_URL}/api/books/${tempBook.id}`, {
    method: "DELETE",
    headers: authHeaders,
  });
  assert.equal(delTempRes.status, 200);
  console.log("✓ Uncirculated book successfully deleted.");

  // 17. AI Intelligence query endpoint
  console.log("\n[Test 16] AI Intelligence: Natural Language Query & Recommendations...");
  // 16a. Unauthenticated check
  const unauthAi = await fetch(`${BASE_URL}/api/ai/query`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ query: "Available books" }),
  });
  assert.equal(unauthAi.status, 401);

  // 16b. Empty query validation
  const emptyAi = await fetch(`${BASE_URL}/api/ai/query`, {
    method: "POST",
    headers: authHeaders,
    body: JSON.stringify({ query: "" }),
  });
  assert.equal(emptyAi.status, 400);

  // 16c. Realistic query: catalog availability
  const aiQueryRes = await fetch(`${BASE_URL}/api/ai/query`, {
    method: "POST",
    headers: authHeaders,
    body: JSON.stringify({ query: "Which books on Python or programming are currently available?" }),
  });
  assert.equal(aiQueryRes.status, 200);
  const aiData = await aiQueryRes.json();
  assert.ok(aiData.answer && typeof aiData.answer === "string");
  assert.ok(Array.isArray(aiData.referencedBooks));
  assert.ok(["gemini", "local_intelligence"].includes(aiData.source));
  console.log("✓ AI query answered successfully via [" + aiData.source + "]. Referenced books count:", aiData.referencedBooks.length);

  // 16d. PWA assets verification
  console.log("\n[Test 17] PWA Compliance: Manifest and Service Worker Assets...");
  const manifestRes = await fetch(`${BASE_URL}/manifest.webmanifest`);
  assert.equal(manifestRes.status, 200);
  const manifestJson = await manifestRes.json();
  assert.equal(manifestJson.name, "LIBRA | Digital Library Management System");
  assert.equal(manifestJson.display, "standalone");
  assert.ok(manifestJson.icons.length >= 2);

  const swRes = await fetch(`${BASE_URL}/sw.js`);
  assert.equal(swRes.status, 200);
  console.log("✓ PWA Manifest and Service Worker validated.");

  console.log("\n🎉 ALL 17 AUTOMATED TESTS PASSED SUCCESSFULLY! 🎉\n");
}

runTests().catch((err) => {
  console.error("\n❌ Test Suite Failed:", err);
  process.exit(1);
});
